finality-provider/service/fp_instance.go

```
@@ -471,15 +583,150 @@ func (fp *FinalityProviderInstance) checkBlockFinalization(height uint64) (bool,
+			return nil, nil
+		}
+	}
+}
+
+// CommitPubRand generates a list of Schnorr rand pairs,
+// commits the public randomness for the managed finality providers,
+// and save the randomness pair to DB
+func (fp *FinalityProviderInstance) CommitPubRand(tipHeight uint64) (*types.TxResponse, error) {
+	lastCommittedHeight, err := fp.GetLastCommittedHeight()
+	if err != nil {
+		return nil, err
+	}
+
+	var startHeight uint64
+	if lastCommittedHeight == uint64(0) {
+		// the finality-provider has never submitted public rand before
+		startHeight = tipHeight + 1
+	} else if lastCommittedHeight < fp.cfg.MinRandHeightGap+tipHeight {
+		// (should not use subtraction because they are in the type of uint64)
+		// we are running out of the randomness
+		startHeight = lastCommittedHeight + 1
+	} else {
+		fp.logger.Debug(
+			"the finality-provider has sufficient public randomness, skip committing more",
+			zap.String("pk", fp.GetBtcPkHex()),
+			zap.Uint64("block_height", tipHeight),
+			zap.Uint64("last_committed_height", lastCommittedHeight),
+		)
+		return nil, nil
+	}
+
+	// generate a list of Schnorr randomness pairs
+	// NOTE: currently, calling this will create and save a list of randomness
+	// in case of failure, randomness that has been created will be overwritten
+	// for safety reason as the same randomness must not be used twice
+	// TODO: should consider an implementation that deterministically create
+	//  randomness without saving it
+	pubRandList, err := fp.createPubRandList(startHeight, fp.cfg.NumPubRand)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate randomness: %w", err)
+	}
+	numPubRand := uint64(len(pubRandList))
+
+	// generate commitment and proof for each public randomness
+	commitment, proofList := types.CommitPubRandList(pubRandList)
+
+	// store them to database
+	if err := fp.pubRandState.AddPubRandProofList(pubRandList, proofList); err != nil {
```

Not sure we should return an error if the error is `ErrDuplicatePubRand` as it could be inconsistency between the remote committed pub rand and ones in the store